src/main/java/org/terasology/thirst/ThirstAuthoritySystem.java

```diff
@@ -42,16 +49,62 @@
  * This authority system handles drink consumption by various entities.
  */
 @RegisterSystem(value = RegisterMode.AUTHORITY)
-public class ThirstAuthoritySystem extends BaseComponentSystem {
+public class ThirstAuthoritySystem extends BaseComponentSystem{
     private static final Logger logger = LoggerFactory.getLogger(ThirstAuthoritySystem.class);
     @In
     private EntityManager entityManager;
     @In
     private InventoryManager inventoryManager;
     @In
+    private PrefabManager prefabManager;
+    @In
     private Time time;
+    @In
+    private DelayManager delayManager;
 
     private boolean destroyDrink = false;
+    private static long lastTick = 0;
+    /**
+     * The interval (in milliseconds) at which healthDecreaseAmount (thirstComponent) is applied to the component.
+     */
+    private int healthDecreaseInterval = 30000;
+    public static final String THIRST_DAMAGE_ACTION_ID = "Thirst Damage";
+
+    /**
+     * Deals a unit of thirst damage to the character.
+     */
+    @ReceiveEvent
+    public void onPeriodicActionTriggered(PeriodicActionTriggeredEvent event, EntityRef entity_unused) {
+        if (event.getActionId().equals(THIRST_DAMAGE_ACTION_ID)) {
+            for (EntityRef entity : entityManager.getEntitiesWith(ThirstComponent.class, AliveCharacterComponent.class)) {
+                ThirstComponent thirst = entity.getComponent(ThirstComponent.class);
+
+                // Check to see if health should be decreased
+                if ((ThirstUtils.getThirstForEntity(entity) < thirst.healthLossThreshold)) {
+                    Prefab thirstDamagePrefab = prefabManager.getPrefab("thirst:thirstDamage");
+                    entity.send(new DoDamageEvent(thirst.healthDecreaseAmount, thirstDamagePrefab));
+                    entity.saveComponent(thirst);
+                    logger.info("onPeriodicActionTriggered method called for character");
+                }
+            }
+        }
+    }
+
+    /**
+     * Cancels the BeforeHealEvent for an entity if their thirst level is lower than the healthLossThreshold.
+     *
+     * @param event  The BeforeHealEvent, called before an entity is about to be healed.
+     * @param entity The entity which is being healed.
+     * @param thirst The ThirstComponent object, containing settings for Thirst.
+     */
+    @ReceiveEvent
+    public void onHealthRegen(BeforeHealEvent event, EntityRef entity,
```

Comment:
I have added this to remove the health regeneration effect when thirst is below threshold

Reply:
Yup that looks correct.  
The `Before` events are usually added in order to allow other systems to cancel of modify them, such as in this case